File: tests/fixed-length-record-test.py
import test
import sys
import re

import logging
logging.basicConfig(stream=sys.stderr, level=logging.DEBUG)

script = 'challenges/fixed-length-record.py'
CPATH= 'content/textfiles';

# get the correct output
def correctOutput(input):
  inputPath= input[0]
  fn= (input[1] + "                ")[0:16]
  ln= (input[2] + "                ")[0:16]
  newBday= (input[3] + '        ')[0:8]
  filecontent= open(inputPath, 'r').read()
  myOutput= re.sub(r''+fn+ln+'........', fn+ln+newBday, filecontent);
  return myOutput

# validation function that the test harness calls
def validate(input, internalTestOutput, studentTestOutput):
  studentOutput= open(input[0]).read()
  return studentOutput == internalTestOutput
  

# clear the output file to the default values
open('/tmp/fixed1','w').write(open(CPATH + '/fixed-length.txt', 'r').read())
open('/tmp/fixed2','w').write(open(CPATH + '/fixed-length.txt', 'r').read())
open('/tmp/fixed3','w').write(open(CPATH + '/poem.txt', 'r').read())

test.test(script, ['/tmp/fixed1', 'Adam', 'Smith', '11111900'], correctOutput, _validate=validate),
test.test(script, ['/tmp/fixed2', 'Adam', 'Smithers', '00000000'], correctOutput, _validate=validate),
test.test(script, ['/tmp/fixed3', 'Adam', 'Smith', '11111111'], correctOutput, _validate=validate);

print ("Well done!")

File: tests/fixed-length-record-test.js
var fs= require('fs');
var test = require('./test-fw.js');

function padString(text, padCount){
  var pad= "";
  for(var i=0; i < padCount; i++) {
    pad += " ";
  }
  return (text + pad).substring(0, padCount)
}

function buildTest(pathi, patho, firstname, lastname, bday){
  var safeOutputPath= '/tmp/' + patho.replace(/\W/g, '')
  
  var startingData= fs.readFileSync(pathi, 'utf8')
  fs.writeFileSync(safeOutputPath, startingData, 'utf8')
  
  var pf= padString(firstname, 16)
  var pl= padString(lastname, 16)
  var re= new RegExp(pf+pl+"........")
	var correctOutput= startingData.replace(re, pf+pl+bday)
  
	return {
		inputs: [safeOutputPath, firstname, lastname, bday],
		outputs: [],
    validate: function(i){
      //console.info("VAL: " + i.inputs[0])
      //console.info(i)
      var actualOutput= fs.readFileSync(i.inputs[0], "utf8");
      var resultAsExpected= correctOutput == actualOutput;
      if(!resultAsExpected){
        console.info("ACTUAL\n" + actualOutput)
        console.info("CORRECT\n" + correctOutput)
      }
      return resultAsExpected;
		}
	}
}

var script = 'challenges/fixed-length-record.js';

var cpath= 'content/textfiles';

test.tests(script, [
	buildTest(cpath+'/fixed-length.txt', 'fixed1', 'Adam', 'Smith', '12121912'),
	buildTest(cpath+'/poem.txt', 'fixed2', 'Adam', 'Smith', '11111111')
]);
